File: app/src/main/java/com/todoroo/astrid/service/TaskDeleter.java
package com.todoroo.astrid.service;

import static org.tasks.db.DbUtils.collect;
import static org.tasks.db.QueryUtils.removeOrder;
import static org.tasks.db.QueryUtils.showHiddenAndCompleted;

import com.google.common.collect.ImmutableList;
import com.todoroo.astrid.api.Filter;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javax.inject.Inject;
import org.tasks.LocalBroadcastManager;
import org.tasks.data.CaldavAccount;
import org.tasks.data.CaldavCalendar;
import org.tasks.data.DeletionDao;
import org.tasks.data.GoogleTaskAccount;
import org.tasks.data.GoogleTaskDao;
import org.tasks.data.GoogleTaskList;
import org.tasks.data.TaskContainer;
import org.tasks.data.TaskListQuery;
import org.tasks.jobs.WorkManager;
import org.tasks.preferences.Preferences;

public class TaskDeleter {

  private final WorkManager workManager;
  private final TaskDao taskDao;
  private final LocalBroadcastManager localBroadcastManager;
  private final GoogleTaskDao googleTaskDao;
  private final Preferences preferences;
  private final DeletionDao deletionDao;

  @Inject
  public TaskDeleter(
      DeletionDao deletionDao,
      WorkManager workManager,
      TaskDao taskDao,
      LocalBroadcastManager localBroadcastManager,
      GoogleTaskDao googleTaskDao,
      Preferences preferences) {
    this.deletionDao = deletionDao;
    this.workManager = workManager;
    this.taskDao = taskDao;
    this.localBroadcastManager = localBroadcastManager;
    this.googleTaskDao = googleTaskDao;
    this.preferences = preferences;
  }

  public int purgeDeleted() {
    List<Long> deleted = deletionDao.getDeleted();
    deletionDao.delete(deleted);
    return deleted.size();
  }

  public void markDeleted(Task item) {
    markDeleted(ImmutableList.of(item.getId()));
  }

  public List<Task> markDeleted(List<Long> taskIds) {
    Set<Long> ids = new HashSet<>(taskIds);
    ids.addAll(collect(taskIds, googleTaskDao::getChildren));
    ids.addAll(collect(taskIds, taskDao::getChildren));
    deletionDao.markDeleted(ids);
    workManager.cleanup(ids);
    workManager.sync(false);
    localBroadcastManager.broadcastRefresh();
    return collect(ids, taskDao::fetch);
  }

  public void delete(Task task) {
    delete(task.getId());
  }

  public void delete(Long task) {
    delete(ImmutableList.of(task));
  }

  public void delete(List<Long> tasks) {
    deletionDao.delete(tasks);
    workManager.cleanup(tasks);
    localBroadcastManager.broadcastRefresh();
  }

  public int clearCompleted(Filter filter) {
    List<Long> completed = new ArrayList<>();
    Filter deleteFilter = new Filter(null, null);
    deleteFilter.setFilterQueryOverride(
        removeOrder(showHiddenAndCompleted(filter.getOriginalSqlQuery())));
    List<TaskContainer> tasks =
        taskDao.fetchTasks(
            (includeGoogleSubtasks, includeCaldavSubtasks) ->
                TaskListQuery.getQuery(
                    preferences, deleteFilter, includeGoogleSubtasks, includeCaldavSubtasks));
    for (TaskContainer task : tasks) {
      if (task.isCompleted()) {
        completed.add(task.getId());
      }
    }
    markDeleted(completed);
    return completed.size();
  }

  public void delete(GoogleTaskList googleTaskList) {
    List<Long> ids = deletionDao.delete(googleTaskList);
    workManager.cleanup(ids);
    localBroadcastManager.broadcastRefresh();
    localBroadcastManager.broadcastRefreshList();
  }

  public void delete(GoogleTaskAccount googleTaskAccount) {
    List<Long> ids = deletionDao.delete(googleTaskAccount);
    workManager.cleanup(ids);
    localBroadcastManager.broadcastRefresh();
    localBroadcastManager.broadcastRefreshList();
  }

  public void delete(CaldavCalendar caldavCalendar) {
    List<Long> ids = deletionDao.delete(caldavCalendar);
    workManager.cleanup(ids);
    localBroadcastManager.broadcastRefresh();
    localBroadcastManager.broadcastRefreshList();
  }

  public void delete(CaldavAccount caldavAccount) {
    List<Long> ids = deletionDao.delete(caldavAccount);
    workManager.cleanup(ids);
    localBroadcastManager.broadcastRefresh();
    localBroadcastManager.broadcastRefreshList();
  }
}

File: app/src/main/java/com/todoroo/astrid/service/TaskDeleter.kt
package com.todoroo.astrid.service

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.tasks.LocalBroadcastManager
import org.tasks.caldav.VtodoCache
import org.tasks.data.dao.DeletionDao
import org.tasks.data.dao.LocationDao
import org.tasks.data.dao.TaskDao
import org.tasks.data.dao.UserActivityDao
import org.tasks.data.db.SuspendDbUtils.chunkedMap
import org.tasks.data.entity.CaldavAccount
import org.tasks.data.entity.CaldavCalendar
import org.tasks.data.entity.Task
import org.tasks.data.pictureUri
import org.tasks.files.FileHelper
import org.tasks.location.GeofenceApi
import org.tasks.notifications.NotificationManager
import org.tasks.sync.SyncAdapters
import javax.inject.Inject

class TaskDeleter @Inject constructor(
    @ApplicationContext private val context: Context,
    private val deletionDao: DeletionDao,
    private val taskDao: TaskDao,
    private val localBroadcastManager: LocalBroadcastManager,
    private val syncAdapters: SyncAdapters,
    private val vtodoCache: VtodoCache,
    private val notificationManager: NotificationManager,
    private val geofenceApi: GeofenceApi,
    private val userActivityDao: UserActivityDao,
    private val locationDao: LocationDao,
) {

    suspend fun markDeleted(item: Task) = markDeleted(listOf(item.id))

    suspend fun markDeleted(taskIds: List<Long>): List<Task> = withContext(NonCancellable) {
        val ids = taskIds
            .toSet()
            .plus(taskIds.chunkedMap(taskDao::getChildren))
            .let { taskDao.fetch(it.toList()) }
            .filterNot { it.readOnly }
            .map { it.id }
        deletionDao.markDeleted(
            ids = ids,
            cleanup = { cleanup(it) }
        )
        syncAdapters.sync()
        localBroadcastManager.broadcastRefresh()
        taskDao.fetch(ids)
    }

    suspend fun delete(task: Task) = delete(task.id)

    suspend fun delete(task: Long) = delete(listOf(task))

    suspend fun delete(tasks: List<Long>) {
        deletionDao.delete(
            ids = tasks,
            cleanup = { cleanup(it) }
        )
        localBroadcastManager.broadcastRefresh()
    }

    suspend fun delete(list: CaldavCalendar) {
        vtodoCache.delete(list)
        deletionDao.delete(
            caldavCalendar = list,
            cleanup = { cleanup(it) }
        )
        localBroadcastManager.broadcastRefreshList()
    }

    suspend fun delete(account: CaldavAccount) {
        vtodoCache.delete(account)
        deletionDao.delete(
            caldavAccount = account,
            cleanup = { cleanup(it) }
        )
        localBroadcastManager.broadcastRefreshList()
    }

    private suspend fun cleanup(tasks: List<Long>) {
        if (tasks.isEmpty()) {
            return
        }
        notificationManager.cancel(tasks)
        tasks.forEach { task ->
            locationDao.getGeofencesForTask(task).forEach {
                locationDao.delete(it)
                geofenceApi.update(it.place!!)
            }
            userActivityDao.getComments(task).forEach {
                FileHelper.delete(context, it.pictureUri)
                userActivityDao.delete(it)
            }
        }
        coroutineScope {
            launch(Dispatchers.IO) {
                notificationManager.updateTimerNotification()
                deletionDao.purgeDeleted()
            }
        }
    }
}
